### ISSUE TITLE
Documenting the complete set of POs for VDM

### ISSUE THREAD

**Issue by nickbattle** [overturetool/language#33]
As discussed in the LB NM of 15/11/15, we are lacking a definition of the complete set of Proof Obligations that can be associated with a VDM specification. This issue is intended to track the progress of creating such a definition.

**Comment by nickbattle** [comment]
The first task is to identify all of the POs that we can find in tools and the literature. The following sources will identify POs, but there may be more:

* VDMTools and VDMJ/Overture
* The "Proof in VDM" and "Systematic Software Development in VDM" books
* Luis' paper about Kleene/McCarthy logic
* Other papers? (PGL may be able to help)

I suspect the majority of these will relate to VDM-SL, or at least the dialect-neutral aspects of all specs. So there will probably be missing POs for dialect specific areas (eg. for class invariants in VDM++). There are also problems with the POs for constructs that appear within operation bodies, so new work may need to be done here too. And we should be aware of the difference between POs that assume McCarthy logic and those that follow a pure Kleene form.

**Comment by leouk** [comment]
I strongly agree with the need for such documentation. Preferably in a formal manner (i.e. a VDM spec or other formal vehicle) to avoid any issue/confusion.

**Comment by nickbattle** [comment]
So the first source is Cliff's "Systematic Software Development in VDM". There are POs defined in Appendix E. But there are only four. Two are satisfiability POs, and two relate to data reification and operation decomposition.

So we need to be clear which aspect of verification POs relate to - either self-consistency, or specification refinement. There may be other aspects as we go on.

**Comment by nickbattle** [comment]
The "Proof in VDM" book mentions satisfiability obligations, but from a very quick skim-read, they are introduced along with the rest of the text rather than being listed explicitly as a set of independent rules in an Appendix. So there are obligations from this source, but we may need to do some work to extract them.

**Comment by JohnFitzgerald** [comment]
You should find all the necessary POs defined in the Proof in VDM book in a bit more detail than in Cliff’s books (because we build up pall the foundation theories). There are also some additional POs, e.g. totality of preconditions. Let me know if you’re having difficulty finding them.

Cheers
John

**Comment by leouk** [comment]
Hi John/Nick,

Many thanks for the pointer. Do you have a copy I could have look of Proof in VDM please?

Nick, I know the POs from Cliff’s book: they are “standard” in the sense that you can test operations for satisfiability and data refinement.

My surprise (from our conversations) was that Overture generate a myriad of POs, many of which throw both of us away as in why they are there.
Some look plausible, others suspicious or incomplete. Others bring to question notions of how VDM works (or should work, or works within Overture).

For instance, in the we were working on, the assumption that functions were referentially transparent was not quite right (and surprised me).
Say, if a boolean-valued function talking about “A and B” short circuits in its result if A is false, then conjunction is no longer commutative, hence
the POs we were puzzling about should indeed be there to ensure functions are pure. On the other hand, the animation aspect of Overture or
else the POG has a problem, yet I couldn’t figure out which.

That’s why I was looking for a list of POs per construct (e.g. what are the production rules the POG has for each VDM construct in Overture?).

best,
Leo

**Comment by nickbattle** [comment]
Leo, there is a "scan" of the Proof in VDM book here: http://overturetool.org/publications/books/proof-in-vdm/ProofinVDM.pdf

Unfortunately we can't search the scanned pages for words like "obligation". John, I don't suppose there is any surviving source material for the book? A tex file perhaps or similar?

**Comment by JohnFitzgerald** [comment]
Regrettably, the tapes containing the source of the proof book had to be sent to the national Physical Laboratory for reading – only they had the old kit. Apparently, they stopped trying when smoke came out of the machine!

**Comment by ldcouto** [comment]
There are many ocr for pdf services. Might be worth a try.

**Comment by peterwvj** [comment]
When I open it in chrome I can search the pdf

**Comment by leouk** [comment]
Thanks for that :-)… I will have a look.
L

**Comment by ldcouto** [comment]
You can see how cleverly I worked on this. I googled for OCR but never actually opened the PDF :)

A while back, I remember starting work on a PO reference guide but it never went very far. I can probably find a pdf of it, if people are interested  (it's very WIP).

**Comment by nickbattle** [comment]
Yes please Luis!

**Comment by ldcouto** [comment]
Here it is, Nick. Note the third author :)

[pogref.pdf](https://github.com/overturetool/language/files/39966/pogref.pdf)

Sidenote: during my PhD I've done a lot of work on the POG so I guess I would like to be a part of this going forward. But I'm not sure if I should (could?) join the LB for that.

**Comment by nickbattle** [comment]
The LB is able to consult area experts, or anyone we think can help. The group is only closed when it comes to voting about what to do with RMs. So you'd be very welcome to be involved in this task!

**Comment by ldcouto** [comment]
Hmm... Sorry, but I don't like the idea of being a consulting expert.

**Comment by nickbattle** [comment]
What would you prefer to be called then? :)

**Comment by peterwvj** [comment]
If you prefer, you can also run for election. Right now there are four seats free and it would be useful to  include people with a background similar to yours.

**Comment by ldcouto** [comment]
Nick, I'm not so shallow that I care about the name. I just think that way I won't have a stake in the project.

I guess I could run for election (who would ever vote for me?). I am looking for a way to stay involved in the community once I leave AU.

**Comment by nickbattle** [comment]
You'd make a fine LB member, Luis. We only have to vote if there are too many people putting themselves forward. Currently, the LB should have seven members (up from the original five).

**Comment by peterwvj** [comment]
It would be great if you decided to run for election  and I'm happy that Leo is already doing it. I really hope that the 2016 LB will include at least a few new members.

**Comment by ldcouto** [comment]
Why not? I'll do it.

**Comment by peterwvj** [comment]
What do you mean "why not" ?

**Comment by ldcouto** [comment]
It's an expression. Like "Yeah, sure".

... isn't it?

**Comment by peterwvj** [comment]
I would have to consult a native speaker... :)

**Comment by nickbattle** [comment]
It's a little "conversational", and so unusual to read in a thread like this. But yes, "why not?" is a rhetorical question, meaning "I can't think of a reason why not".

Now can we get back to POs? :-)

**Comment by ldcouto** [comment]
Yes, we can. Have you skimmed the reference guide? I don't think there's a
ton of content on it but maybe it's a good template for the document we end
up producing.

I also remember looking into state POs as part of COMPASS but I can't find
anything on it. I'll ask Victor next week. He was a part of it too.

**Comment by nickbattle** [comment]
Yes, I think it's at least the start of a document that we should be aiming at. A few points:

* The introduction mentions internal consistency checks, so this is not including refinement POs (and the tools don't produce them). But while we're doing this exercise, I think we should capture them too.
* It's interesting that it has separated the basic POs from the contexts that lead up to them. As long as contexts are always the same for any PO at a position in a spec, that is probably a sensible distinction, and it doesn't clutter the POs with context distractions.
* I would like to find some sort of rational categorization for the POs. That way, we may be able to spot ones we've missed more easily. For example, many of the POs are concerned with partial operators, but others are concerned with pre/post satisfiability or invariant preservation. So I see these as broad categories. In my mind, there is a table with expression and statement type rows, and PO category columns, sparsely populated with PO logic where a statement or expression results in a PO of each category. But maybe that's too simple.
* It does mention the Kleene/McCarthy logic question, which we do have to tackle somehow. Not sure how though!

**Comment by nickbattle** [comment]
Perhaps there could also be a "column" for the context that a particular expression or statement creates as well? That might work for expressions, though I think things are will be more complicated with statements, so perhaps that wouldn't work.

**Comment by ldcouto** [comment]
As part of the COMPASS work, I remember sorting the POs into 3 or 4
categories. [insert rant about never writing the COMPASS POG paper]. I do
like the idea of some kind of summary table.

On LPF, I wrote a POG prototype that computes LPF POs --- it's actually not
that much work. But to me, the question is bigger than that. We need to
decide if the POG will support LPF or not. If it does, what about the
interpreter? No to mention a hypothetical theorem prover. Others may know
more but I think LPF support is very poor in that area.

I think, for now, we should start with a reference for the POs we have and
explain that there is actually a disconnect between the POs produced and
the semantics of the language. That disconnect is related to the
implementation of the interpreter. Bridging that gap can be seen as future
work.

**Comment by nickbattle** [comment]
I tried using Chrome (actually, Chromium, but a recent version) to search the "Proof in VDM" PDF file, but it would not work for me. Perhaps this is a Windows feature? I thought that it would be useful to try to OCR the images anyway, and after a bit of investigation I managed to produce the following - a ZIP of the individual page images (as Portable Bit Maps, pdms), plus the OCR text of the same.

The OCR had a lot of fun with some of the unusual symbols, as you can imagine! But given we have the corresponding page images it should be easy to find the actual expressions in each case. The actual text conversion seems very accurate.

[ProofinVDM.zip](https://github.com/overturetool/language/files/76630/ProofinVDM.zip)

**Comment by peterwvj** [comment]
I'm on Linux (Ubuntu Gnome) and it works for me so it might be because you are using Chromium (I'm not so sure how it's different from Chrome).

**Comment by ldcouto** [comment]
Chromium is slightly different from Chrome so that may be it.

**Comment by nickbattle** [comment]
Yes, Chromium has only open source components apparently, but I couldn't find anything that said that the PDF viewer was different.

It wasn't hard to get the document OCR'd: "pdfimages" extracts the pages individually and "tesseract" does the OCR. I think it will be useful to have the entire text of the Proof In VDM book available, even if the mathematical notation is missing.

**Comment by nickbattle** [comment]
This issue has been dormant for a while, and an action was created on the LB meeting of the 17th Feb to try to kickstart it by creating a plan of the way forward.

Re-reading the records above, a lot of the base work has already been started. We have a reasonable list of sources of POs that have been documented, and "Proof In VDM" has been OCR scanned to allow extraction of some of them from there.

There is also the skeleton of an idea for the structure of a document, influenced by Luis' existing paper and from what we know about consistency/refinement POs and contexts.

So I think the obvious next step is to try to draft something, so that we can settle on a document structure that captures what we need to say. I have some ideas of what that can be like.

After that, the details need to be filled in, capturing all the POs and contexts that we know about. If the document structure is right, we may be able to identify missing obligations at this point (I'm hoping that groups of related expressions or statements will produce similar POs, so we may be able to identify gaps).

The tools (POGs) can then be tested against the document. A test suite should be developed in parallel with the document, and included in the tools' build systems to verify that they are generating all the correct POs (at least as far as consistency POs are concerned; tools don't yet generate refinement POs).

There will then be a harder stage when we try to resolve the conflict between the McCarthy logic in the tools and the Kleene/LPF logic in pure VDM and its theory. This may seem fundamental, but if we try to tackle that first, we might get bogged down and never produce anything. So the steps above will document what the tools (should!) produce, and can explain the issue with the logic and the reason why the interpreters work as they do.

**Comment by nickbattle** [comment]
A couple of updates that are relevant to this working group:

1. The QuickCheck tool has come a long way and is able to find counterexamples for some POs, as well as indicating that some are probably provable. The tool is integrated with the VDM-VSCode extension, as well as command line VDMJ.

2. A lot of work has been done on the creation of POs for explicit operations. This used to be rudimentary, but now POs are generated for many cases in VDM-SL. A few things remain unsupported, such as exception handling, and support for VDM++ and VDM-RT is very simple because of the complex state they maintain.

There are a couple of Overture workshop papers on these aspects, and there are some VDMJ wiki pages being maintained, which show the latest features:

https://arxiv.org/abs/2410.02046
https://arxiv.org/abs/2506.12858
https://github.com/nickbattle/vdmj/wiki/Using-QuickCheck
https://github.com/nickbattle/vdmj/wiki/Operation-Obligations
https://github.com/nickbattle/vdmj/wiki/Loop-Invariant-Annotations
